File: ScorePulse/Models/Score.swift
import Foundation

/// Represents a complete musical score
struct Score: Codable, Identifiable {
    let id: UUID
    let title: String
    let composer: String
    let defaultTempo: Int
    let tempoChanges: [TempoChange]
    let rehearsalMarks: [RehearsalMark]
    let bars: [Bar]  // Only stores bars with time signature changes
    let totalBars: Int
    
    init(id: UUID = UUID(), title: String, composer: String, defaultTempo: Int,
         tempoChanges: [TempoChange] = [], rehearsalMarks: [RehearsalMark] = [],
         bars: [Bar], totalBars: Int) {
        self.id = id
        self.title = title
        self.composer = composer
        self.defaultTempo = defaultTempo
        self.tempoChanges = tempoChanges.sorted { $0.bar < $1.bar }
        self.rehearsalMarks = rehearsalMarks.sorted { $0.bar < $1.bar }
        self.bars = bars.sorted { $0.number < $1.number }
        self.totalBars = totalBars
    }
    
    /// Get the time signature for a specific bar number
    func timeSignature(at barNumber: Int) -> TimeSignature {
        // Find the most recent bar with a time signature at or before this bar
        let relevantBars = bars.filter { $0.number <= barNumber }
        if let mostRecent = relevantBars.last {
            return mostRecent.timeSignature
        }
        // Default to 4/4 if no time signature found
        return .fourFour
    }
    
    /// Get the base tempo for a specific bar number (ignores transitions)
    func tempo(at barNumber: Int) -> Int {
        // Find the most recent tempo change with an actual tempo at or before this bar
        let relevantChanges = tempoChanges.filter { $0.bar <= barNumber && $0.tempo != nil }
        if let mostRecent = relevantChanges.last, let tempo = mostRecent.tempo {
            return tempo
        }
        return defaultTempo
    }
    
    /// Get the tempo for a specific bar and beat position, accounting for gradual transitions
    /// - Parameters:
    ///   - barNumber: The bar number
    ///   - beatProgress: Progress through the bar (0.0 to 1.0), where 0.0 is start and 1.0 is end
    /// - Returns: The interpolated tempo at this position
    func tempo(at barNumber: Int, beatProgress: Double) -> Int {
        // Check if we're in a transition range (includes both single and multi-bar transitions)
        if let (startBar, endBar, startTempo, endTempo) = findTransitionRange(containing: barNumber) {
            // Calculate overall progress through the transition range
            let totalBars = endBar - startBar
            let barsCompleted = barNumber - startBar
            let overallProgress = (Double(barsCompleted) + beatProgress) / Double(totalBars)
            
            let interpolated = Double(startTempo) + (Double(endTempo - startTempo) * overallProgress)
            return Int(interpolated)
        }
        
        // No transition, return base tempo
        return tempo(at: barNumber)
    }
    
    /// Find the tempo before a transition starts
    private func tempoBeforeTransition(at barNumber: Int) -> Int {
        // Look for the most recent bar with an actual tempo before this bar
        let previousTempoChanges = tempoChanges.filter { $0.bar < barNumber && $0.tempo != nil }
        if let previous = previousTempoChanges.last, let tempo = previous.tempo {
            return tempo
        }
        return defaultTempo
    }
    
    /// Find if a bar is within a transition range (one or more acc/rit bars leading to a tempo)
    /// - Returns: (startBar, endBar, startTempo, endTempo) or nil if not in a transition
    private func findTransitionRange(containing barNumber: Int) -> (Int, Int, Int, Int)? {
        // Find all transition markers
        let transitionMarkers = tempoChanges.filter { $0.transition != .none }
        
        for marker in transitionMarkers {
            // Find the end of this transition (next bar with a tempo)
            guard let endChange = tempoChanges.first(where: { $0.bar > marker.bar && $0.tempo != nil }),
                  let endTempo = endChange.tempo else {
                continue
            }
            
            let startBar = marker.bar
            let endBar = endChange.bar
            
            // Check if barNumber is within this range
            if barNumber >= startBar && barNumber < endBar {
                let startTempo = tempoBeforeTransition(at: startBar)
                return (startBar, endBar, startTempo, endTempo)
            }
        }
        
        return nil
    }
    
    /// Check if a bar is in a tempo transition
    func isInTransition(at barNumber: Int) -> Bool {
        // Direct transition marker on this bar
        if tempoChanges.contains(where: { $0.bar == barNumber && $0.transition != .none }) {
            return true
        }
        // Part of a multi-bar transition range
        return findTransitionRange(containing: barNumber) != nil
    }
    
    /// Get the tempo marking for a specific bar number
    func tempoMarking(at barNumber: Int) -> String? {
        let relevantChanges = tempoChanges.filter { $0.bar <= barNumber }
        return relevantChanges.last?.marking
    }
    
    /// Get rehearsal mark at a specific bar, if any
    func rehearsalMark(at barNumber: Int) -> RehearsalMark? {
        rehearsalMarks.first { $0.bar == barNumber }
    }
    
    /// Get the next rehearsal mark after a given bar
    func nextRehearsalMark(after barNumber: Int) -> RehearsalMark? {
        rehearsalMarks.first { $0.bar > barNumber }
    }
    
    /// Get the previous rehearsal mark before a given bar
    func previousRehearsalMark(before barNumber: Int) -> RehearsalMark? {
        rehearsalMarks.reversed().first { $0.bar < barNumber }
    }
}

/// Container for loading multiple scores from JSON
struct ScoreCollection: Codable {
    let scores: [Score]
}

File: ScorePulseAndroid/app/src/main/java/com/scorepulse/model/Score.kt
package com.scorepulse.model

import kotlinx.serialization.Serializable
import java.util.UUID

/**
 * Represents a complete musical score
 */
@Serializable
data class Score(
    @Serializable(with = UUIDSerializer::class)
    val id: UUID = UUID.randomUUID(),
    val title: String,
    val composer: String,
    val defaultTempo: Int,
    val tempoChanges: List<TempoChange> = emptyList(),
    val rehearsalMarks: List<RehearsalMark> = emptyList(),
    val bars: List<Bar>,
    val totalBars: Int
) {
    private val sortedTempoChanges = tempoChanges.sortedBy { it.bar }
    private val sortedRehearsalMarks = rehearsalMarks.sortedBy { it.bar }
    private val sortedBars = bars.sortedBy { it.number }

    /**
     * Get the time signature for a specific bar number
     */
    fun timeSignature(barNumber: Int): TimeSignature {
        val relevantBars = sortedBars.filter { it.number <= barNumber }
        return relevantBars.lastOrNull()?.timeSignature ?: TimeSignature.FOUR_FOUR
    }

    /**
     * Get the base tempo for a specific bar number (ignores transitions)
     */
    fun tempo(barNumber: Int): Int {
        val relevantChanges = sortedTempoChanges.filter { it.bar <= barNumber && it.tempo != null }
        return relevantChanges.lastOrNull()?.tempo ?: defaultTempo
    }

    /**
     * Get the tempo for a specific bar and beat position, accounting for gradual transitions
     */
    fun tempo(barNumber: Int, beatProgress: Double): Int {
        val transitionRange = findTransitionRange(barNumber)
        if (transitionRange != null) {
            val (startBar, endBar, startTempo, endTempo) = transitionRange
            val totalBars = endBar - startBar
            val barsCompleted = barNumber - startBar
            val overallProgress = (barsCompleted + beatProgress) / totalBars
            return (startTempo + (endTempo - startTempo) * overallProgress).toInt()
        }
        return tempo(barNumber)
    }

    private fun tempoBeforeTransition(barNumber: Int): Int {
        val previousChanges = sortedTempoChanges.filter { it.bar < barNumber && it.tempo != null }
        return previousChanges.lastOrNull()?.tempo ?: defaultTempo
    }

    private fun findTransitionRange(barNumber: Int): TransitionRange? {
        val transitionMarkers = sortedTempoChanges.filter { it.transition != TempoTransition.none }

        for (marker in transitionMarkers) {
            val endChange = sortedTempoChanges.firstOrNull { it.bar > marker.bar && it.tempo != null }
                ?: continue
            val endTempo = endChange.tempo ?: continue

            val startBar = marker.bar
            val endBar = endChange.bar

            if (barNumber >= startBar && barNumber < endBar) {
                val startTempo = tempoBeforeTransition(startBar)
                return TransitionRange(startBar, endBar, startTempo, endTempo)
            }
        }
        return null
    }

    /**
     * Check if a bar is in a tempo transition
     */
    fun isInTransition(barNumber: Int): Boolean {
        if (sortedTempoChanges.any { it.bar == barNumber && it.transition != TempoTransition.none }) {
            return true
        }
        return findTransitionRange(barNumber) != null
    }

    /**
     * Get the tempo marking for a specific bar number
     */
    fun tempoMarking(barNumber: Int): String? {
        val relevantChanges = sortedTempoChanges.filter { it.bar <= barNumber }
        return relevantChanges.lastOrNull()?.marking
    }

    /**
     * Get rehearsal mark at a specific bar, if any
     */
    fun rehearsalMark(barNumber: Int): RehearsalMark? {
        return sortedRehearsalMarks.firstOrNull { it.bar == barNumber }
    }

    /**
     * Get the next rehearsal mark after a given bar
     */
    fun nextRehearsalMark(barNumber: Int): RehearsalMark? {
        return sortedRehearsalMarks.firstOrNull { it.bar > barNumber }
    }

    /**
     * Get the previous rehearsal mark before a given bar
     */
    fun previousRehearsalMark(barNumber: Int): RehearsalMark? {
        return sortedRehearsalMarks.lastOrNull { it.bar < barNumber }
    }
}

private data class TransitionRange(
    val startBar: Int,
    val endBar: Int,
    val startTempo: Int,
    val endTempo: Int
)

/**
 * Container for loading multiple scores from JSON
 */
@Serializable
data class ScoreCollection(
    val scores: List<Score>
)
